bundle.go

```diff
@@ -38,59 +41,76 @@ func unarchiveBundle(bundleURL, targetDir string) (err error) {
 		return fmt.Errorf("invalid path, expecting it to begin with: %s", flutterPath)
 	}
 
-	if err := runRequest(bundleURL, targetDir); err != nil {
-		return fmt.Errorf("failed to download bundle, error: %s", err)
+	bundleTarPth, err := downloadBundle(bundleURL)
+	if err != nil {
+		return err
+	}
+
+	if err := unarchiveBundle(bundleTarPth, targetDir); err != nil {
+		return err
 	}
 	return nil
 }
 
-func runRequest(bundleURL string, targetDir string) (err error) {
-	return retry.Times(2).Wait(5 * time.Second).Try(func(attempt uint) error {
+func downloadBundle(bundleURL string) (string, error) {
+	var resp *http.Response
+	if err := retry.Times(2).Wait(5 * time.Second).Try(func(attempt uint) error {
 		if attempt > 0 {
 			log.TWarnf("%d query attempt failed", attempt)
 		}
 
-		resp, err := http.Get(bundleURL)
-
-		defer func() {
-			if resp.Body != nil {
-				cerr := resp.Body.Close()
-				if err == nil {
-					err = cerr
-					return
-				}
-				log.Warnf("failed to close response body, error: %s", cerr)
-			}
-		}()
-
+		var err error
+		resp, err = http.Get(bundleURL)
 		if err != nil {
 			return err
 		}
 
 		if resp.StatusCode != http.StatusOK {
-			responseBody, err := ioutil.ReadAll(resp.Body)
+			body, err := ioutil.ReadAll(resp.Body)
 			if err != nil {
 				return err
 			}
 
-			return fmt.Errorf("query failed, status code: %d, response body: %s", resp.StatusCode, responseBody)
+			return fmt.Errorf("query failed, status code: %d, response body: %s", resp.StatusCode, body)
 		}
 
-		// using -J to support tar.xz
-		tarCmd, err := command.NewWithParams("tar", "-xJf", "-", "-C", targetDir)
-		if err != nil {
-			return fmt.Errorf("failed to create command, error: %s", err)
-		}
-		tarCmd.SetStdin(resp.Body)
+		return nil
+	}); err != nil {
+		return "", err
+	}
 
-		out, err := tarCmd.RunAndReturnTrimmedCombinedOutput()
-		if err != nil {
-			if errorutil.IsExitStatusError(err) {
-				return fmt.Errorf("tar command failed: %s, out: %s", err, out)
-			}
-			return fmt.Errorf("failed to run tar command, error: %s", err)
+	tmpDir, err := pathutil.NormalizedOSTempDirPath("__flutter-sdk__")
+	if err != nil {
+		return "", err
+	}
+
+	sdkTarPth := filepath.Join(tmpDir, "flutter.tar")
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
```

Comment:
You could open a file and use io.Copy to stream the archive as it is downloading, that would use less memory and time.